### PR TITLE
Detail View Config

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #3526
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/api/metadata.py [hunk 1]
```diff
@@ -71,6 +74,14 @@ def _get_field_properties(self, field: schema.SerializerType, name: str) -> Dict
...
             result["items"] = self._get_field_properties(field.child_relation, "")
             result["uniqueItems"] = True
+        elif isinstance(field, NautobotHyperlinkedRelatedField):
+            result["format"] = type_map_obj["format"]
+            model_options = field.queryset.model._meta
+            # Custom Keyword: modelName and appName
+            # This Keyword represents the model name of the uuid model
+            # and appName represents the app_name of the model
+            result["modelName"] = model_options.model_name
+            result["appName"] = model_options.app_label
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:84]: Do `modelName` and `appName`/`appLabel` need to also be added to the `ManyRelatedField` case?

File: nautobot/dcim/api/serializers.py [hunk 2]
```diff
@@ -199,6 +199,18 @@ class Meta:
...
+                "Location": {"fields": ["name", "display", "rack_count", "parent"]},
+                "Management": {"fields": ["status", "prefix_count", "device_count"]},
+                "Others": {"fields": ["tenant", "natural_key_slug", "url"]},
+            },
+            {
+                "Tags": {"fields": ["tags"]},
+                "Comments": {"fields": ["comments"]},
+                "Contact": {"fields": ["contact_name", "contact_phone", "contact_email"]},
+            },
+        ]
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:213]: We need to be sure to document the schema for this attribute in the docs.

File: nautobot/dcim/api/serializers.py [hunk 3]
```diff
@@ -199,6 +199,18 @@ class Meta:
...
+            {
+                "Location": {"fields": ["name", "display", "rack_count", "parent"]},
+                "Management": {"fields": ["status", "prefix_count", "device_count"]},
+                "Others": {"fields": ["tenant", "natural_key_slug", "url"]},
+            },
+            {
+                "Tags": {"fields": ["tags"]},
+                "Comments": {"fields": ["comments"]},
+                "Contact": {"fields": ["contact_name", "contact_phone", "contact_email"]},
+            },
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:212]: Shouldn't at least some of these have `template: list`?

File: nautobot/dcim/api/serializers.py [hunk 4]
```diff
@@ -199,6 +199,18 @@ class Meta:
         list_display_fields = ["name", "status", "parent", "tenant", "description", "tags"]
         # https://www.django-rest-framework.org/api-guide/validators/#optional-fields
         validators = []
+        detail_view_config = [
+            {
+                "Location": {"fields": ["name", "display", "rack_count", "parent"]},
+                "Management": {"fields": ["status", "prefix_count", "device_count"]},
+                "Others": {"fields": ["tenant", "natural_key_slug", "url"]},
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:206]: We should think about whether we can have a pattern for "standard" fields to not need to be explicitly called out in the detail_view_config (i.e. `natural_key_slug`, `url`, `display`, `status` should always be handled consistently, so it's redundant and error-prone to have to call them out in every serializer).

Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py:206]: Excellent idea, but where would these fields be located?
I'm presuming it's typically on Col 1 Group 1?

File: nautobot/dcim/api/serializers.py (hunk 3 above)
Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py:212]: As of now, just the list template can be defined; I don't see the need in explicitly declaring it again.

File: nautobot/core/api/metadata.py [hunk 5]
```diff
@@ -256,11 +268,18 @@ def determine_metadata(self, request, view):
...
             metadata["view_options"] = self.determine_view_options(request, serializer)
 
         return metadata
+
+    def determine_detail_view_schema(self, serializer):
+        view_config = getattr(
+            serializer.Meta,
+            "detail_view_config",
+            SerializerDetailViewConfig(serializer).view_config(),
+        )
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:284]: Not sure i understand

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/api/metadata.py:284]:
```suggestion
        if hasattr(serializer.Meta, "detail_view_config"):
            view_config = serializer.Meta.detail_view_config
        else:
            view_config = SerializerDetailViewConfig(serializer).view_config()
```

File: nautobot/dcim/api/serializers.py (hunk 4 above)
Comment (glennmatthews) [reply nautobot/dcim/api/serializers.py:206]: It's going to depend on the field:

- Status and Display appear in the page header already and probably can be simply ignored in the detail_view_config
- ~natural key slug~ ("composite key") and URL and ID probably should all appear at the end of column 1 group 1.

File: nautobot/dcim/api/serializers.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/dcim/api/serializers.py:212]: OK, just seems asymmetrical to me that `view_config()` returns a list of dicts with `fields` and `template` as keys, while a statically declared `detail_view_config` only contains `fields`?

File: nautobot/ui/src/utils/url.js [hunk 6]
```diff
@@ -2,3 +2,15 @@
 export function uiUrl(apiUrl) {
     return apiUrl.replace("/api/", "/");
 }
+
+/* Build url form app_label, model_name and or pk */
+export function buildUrl(appLabel, modelName, pk = null, isPlugin = false) {
+    let url = `/${appLabel}/${modelName}/`;
+    if (isPlugin) {
+        url = "/plugin" + url;
```
Comment (glennmatthews) [inline nautobot/ui/src/utils/url.js:10]:
```suggestion
        url = "/plugins" + url;
```

File: nautobot/core/api/metadata.py (hunk 1 above)
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:84]: `ManyRelatedField` would recursively call the `_get_field_properties`, which would end up here `elif isinstance(field, drf_serializers.RelatedField):`

File: nautobot/ui/src/components/ObjectDetail/RenderChangeLogsTab.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import useSWR from "swr";
```
Comment (HanlinMiao) [inline nautobot/ui/src/components/ObjectDetail/RenderChangeLogsTab.js:1]: NOICE

File: nautobot/ui/src/components/Table.js [hunk 8]
```diff
@@ -1,48 +1,99 @@
...
+            size="xs"
+            variant="table"
+            onClick={() => alert(`Clicked ${cellContext.row.original.name}!`)}
+        />
+    );
+}
+
 // A standard Nautobot table. This _may_ be beneficial to move into nautobot-ui
-export default function NautobotTable({ data, headers }) {
+export default function NautobotTable({
```
Comment (HanlinMiao) [inline nautobot/ui/src/components/Table.js:59]: Sorry for bothering you with this but can I get some comments on what some of the react hooks are doing?

Comment (timizuoebideri1) [reply nautobot/ui/src/components/Table.js:59]: Im attaching link to the documentation for each hook:

1. [useMemo](https://react.dev/reference/react/useMemo)
2. [useCallback](https://react.dev/reference/react/useCallback)

Its basically a copy paste code from what was in ObjectRetrieve.js initially

File: nautobot/dcim/api/serializers.py (hunk 2 above)
Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py]: Do you have any suggestions about where this document should be added?

File: nautobot/core/testing/api.py [hunk 9]
```diff
@@ -862,6 +862,11 @@ def test_options_returns_expected_choices(self):
                         field_choices.add(field_name)
             else:
                 self.fail(f"Neither PUT nor POST are available actions in: {data['actions']}")
+                
+            print("========")
+            print(set(self.choices_fields))
+            print(field_choices)
+            print("========")
```
Comment (bryanculver) [inline nautobot/core/testing/api.py:869]:
```suggestion
```

File: nautobot/ui/src/components/ObjectDetail/RenderCol.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import { NautobotGrid, NautobotGridItem } from "@nautobot/nautobot-ui";
+
+import RenderGroup from "./RenderGroup";
+
+function RenderCol({ tabData, data, schema }) {
```
Comment (bryanculver) [inline nautobot/ui/src/components/ObjectDetail/RenderCol.js:5]: I think many of these are good enough to be used generically. Maybe `components/render` instead of `components/ObjectDetail`?

File: nautobot/ui/src/views/generic/ObjectRetrieve.js [hunk 11]
```diff
@@ -362,190 +44,45 @@ export default function ObjectRetrieve({ api_url }) {
...
+        tabs: {
+            Location: schemaData.view_options.retrieve,
+        },
+    };
+
+    return (
+        <GenericView objectData={data}>
+            <Box background="white-0" borderRadius="md">
+                <RenderHeader data={data} />
+                <RenderTabs
```
Comment (bryanculver) [inline nautobot/ui/src/views/generic/ObjectRetrieve.js:76]: Much better 👏 

Let's update the names to be more descriptive, something like `RetreiveViewHeader` and `RetreiveViewTabs`.

**Review by bryanculver — COMMENTED** [review]
Loving the direction this is going in multiple ways. Some PR feedback, ensure tests are passing, then I'd say this is GTG.

**Comment by bryanculver** [comment]
@timizuoebideri1 I pulled this down and want to know if you're seeing this in your environment. Looks like `id`, `natural_key_slug`, and `url` are being added multiple times.


image

image

File: nautobot/core/api/utils.py [hunk 12]
```diff
@@ -349,3 +350,71 @@ def return_nested_serializer_data_based_on_depth(serializer, depth, obj, obj_rel
...
+            ]
+
+        Returns:
+            A list representing the view config.
+        """
+        m2m_fields, other_fields = self.get_m2m_and_non_m2m_fields()
+        model_verbose_name = self.serializer.Meta.model._meta.verbose_name
+        return [
+            {
+                model_verbose_name.capitalize(): {
```
Comment (bryanculver) [inline nautobot/core/api/utils.py:415]: `verbose_name` should already properly capitilized: https://github.com/nautobot/nautobot/blob/496ce1000c8c5cb5962dbd58ca2ed190fbeaf028/nautobot/ipam/models.py#L102

Re-capitalizing this is causing the names to appear incorrect.

File: nautobot/core/api/utils.py [hunk 13]
```diff
@@ -386,6 +386,10 @@ def get_m2m_and_non_m2m_fields(self):
 
         return m2m_fields, non_m2m_fields
 
+    def capitalize(self, value: str):
+        """Capitalize value only if not uppercase."""
+        return value if (value.isupper()) else value.capitalize()
```
Comment (bryanculver) [inline nautobot/core/api/utils.py:391]: This won't work as well for cases like `IP Address`

**Review by bryanculver — CHANGES_REQUESTED** [review]

File: nautobot/core/api/utils.py (hunk 13 above)
Comment (glennmatthews) [reply nautobot/core/api/utils.py:391]: Looks like my comment on the commit itself isn't showing up here - should be able to use the existing `bettertitle` function for this purpose.

File: nautobot/dcim/api/serializers.py [hunk 14]
```diff
@@ -199,6 +199,19 @@ class Meta:
         list_display_fields = ["name", "status", "parent", "tenant", "description", "tags"]
         # https://www.django-rest-framework.org/api-guide/validators/#optional-fields
         validators = []
+        # TODO: Add documentation of this schema
+        detail_view_config = [
+            {
+                "Location": {"fields": ["name", "rack_count", "parent"]},
+                "Management": {"fields": ["status", "prefix_count", "device_count"]},
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:206]: `"status"` isn't necessary here right?

Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:206]: why is `rack_count` in `Location` but `prefix_count` and `device_count` in Management? Any specific reasoning?

File: nautobot/dcim/api/serializers.py [hunk 15]
```diff
@@ -199,6 +199,19 @@ class Meta:
...
+                "Others": {"fields": ["tenant"]},
+            },
+            {
+                "Tags": {"fields": ["tags"]},
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:210]: As we move toward more and more models having `tenant` and `tags` fields, it'd be good to have these handled in a standardized way too, I think?

File: nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    const defaultHeaders = schema.view_options.list_display_fields;
+    const tableHeaders = schema.view_options.fields;
+    const tableData = data?.results || [];
+
+    return (
+        <NautobotGrid columns="4">
+            <NautobotGridItem>
+                <Heading display="flex" alignItems="center" gap="5px">
+                    <NtcThumbnailIcon width="25px" height="30px" /> Change Logs
+                </Heading>
```
Comment (glennmatthews) [inline nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js:55]: Not blocking, but it might be useful to create a reusable component for this `<Heading ...><NtcThumbnailIcon/> {title} </Heading>` pattern since you're using it in at least three places in this PR alone.

File: nautobot/ui/src/components/RetrieveViewHeader.js
Comment (glennmatthews) [inline nautobot/ui/src/components/RetrieveViewHeader.js:1]: Should this be `components/Retrieve/ViewHeader.js`? Not clear to me the rationale why some components are in the subdirectory and others aren't.

File: nautobot/ui/src/components/RetrieveViewTabs.js
Comment (glennmatthews) [inline nautobot/ui/src/components/RetrieveViewTabs.js:1]: Same question here - `components/Retrieve/ViewTabs.js`?

File: nautobot/ui/src/components/Table.js
Comment (glennmatthews) [inline nautobot/ui/src/components/Table.js:1]: Not blocking but it would be worth thinking about the name of our `Table` and `TableItem` components, as they're a bit confusing against the `nautobot-ui/Table` component and I don't *think* they're meant as generic components - they have a fairly specific set of applications, don't they?

File: nautobot/core/api/metadata.py [hunk 17]
```diff
@@ -72,41 +76,47 @@ def _get_field_properties(self, field: schema.SerializerType, name: str) -> Dict
...
         return result
 
+    @staticmethod
+    def _filter_fields(all_fields):
+        """
+        Override super._filter_fields to return all fields, including read-only fields,
+        as read-only fields have to be displayed in Detail View.
+        """
+        return tuple((name, field) for name, field in all_fields)
+
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/api/metadata.py:119]: I *think* this results in read-only fields now appearing in the create/edit forms by default, which seems undesirable?

[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:119]: In the creat/edit field we could hide read-only fields

File: nautobot/core/api/metadata.py [hunk 18]
```diff
@@ -256,11 +266,56 @@ def determine_metadata(self, request, view):
...
+                    value["fields"].extend(fields_to_add)
+        return view_config
+
+    def determine_detail_view_schema(self, serializer):
+        """Determine the layout option that would be used for the detail view"""
+        if hasattr(serializer.Meta, "detail_view_config"):
+            view_config = serializer.Meta.detail_view_config
+        else:
+            view_config = SerializerDetailViewConfig(serializer).view_config()
+        return self.restructure_view_config(view_config)
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:321]: According to our discussions, the idea was to leave it as a json schema for the time being.

File: nautobot/dcim/api/serializers.py (hunk 14 above)
Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py:206]: Nope, Could not think of a better grouping,
Illd use the grouping on the legacy UI 👍

File: nautobot/ui/src/components/Table.js
Comment (timizuoebideri1) [reply nautobot/ui/src/components/Table.js:1]: It could be used as a generic Table component, for ChangeLogs, Notes, or in the ListView.

File: nautobot/core/api/metadata.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/api/metadata.py]: I'm not suggesting changing the schema, just how it's implemented in code.

File: nautobot/ui/src/components/Table.js
Comment (glennmatthews) [reply nautobot/ui/src/components/Table.js:1]: So maybe something like `ObjectTable` or `RecordTable`?

File: nautobot/core/api/metadata.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:119]: Nope

[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:119]: It's in the schema

[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:119]: The schema tells you the field type to employ, such as CharField, DynamicSelect, and so on. At that point, you can determine if a field is readonly and hide it if needed.

File: nautobot/core/api/metadata.py [hunk 19]
```diff
@@ -72,41 +77,47 @@ def _get_field_properties(self, field: schema.SerializerType, name: str) -> Dict
...
+            if isinstance(field, drf_serializers.RelatedField):
+                result["uniqueItems"] = True
+                if hasattr(field, "queryset") and hasattr(field.queryset, "model"):
+                    model_options = field.queryset.model._meta
+                    result["required"] = field.required
+                    # Custom Keyword: modelName and appLabel
+                    # This Keyword represents the model name of the uuid model
+                    # and appLabel represents the app_name of the model
+                    result["modelName"] = model_options.model_name
+                    result["appLabel"] = model_options.app_label
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:89]: Do we need to add these keys for the `ManyRelatedField` case above as well?

File: nautobot/core/api/metadata.py [hunk 20]
```diff
@@ -72,41 +77,47 @@ def _get_field_properties(self, field: schema.SerializerType, name: str) -> Dict
...
-            if format_:
+                    result["enum"] = [item[0] for item in enum]
+                    result["enumNames"] = [item[1] for item in enum]
+            if format_ := type_map_obj["format"]:
                 result["format"] = format_
 
-            # Process "readOnly"
-            read_only = type_map_obj["readOnly"]
-            if read_only:
+            if read_only := field.read_only:
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:103]: Why the change from `type_map_obj["readOnly"]` to `field.read_only`?

File: nautobot/core/api/metadata.py [hunk 21]
```diff
@@ -256,11 +267,111 @@ def determine_metadata(self, request, view):
...
+                    Location: {fields: ["name","id","natural_key_slug","url"]},
+                    Others: {fields: ["tenant", "tenant_group"]}
+                },
+                ...
+            ]
+        """
+
+        # TODO(timizuo): Add a standardized way of handling `tenant` and `tags` fields, Possible should be on last items on second col.
+        fields_to_remove = ["natural_key_slug", "url", "display", "status", "id"]
+        fields_to_add = ["id", "natural_key_slug", "url"]
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:306]: This will of course conflict with #3881.

File: nautobot/core/api/metadata.py [hunk 22]
```diff
@@ -256,11 +267,111 @@ def determine_metadata(self, request, view):
...
-                    # "uiSchema": NautobotUiSchemaProcessor(serializer, request.parser_context).get_ui_schema(),
                     "uiSchema": ui_schema,
                 }
             )
 
             metadata["view_options"] = self.determine_view_options(request, serializer)
 
         return metadata
+
+    def restructure_view_config(self, view_config):
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:278]: Could you please add some unit tests for the functionality added in this PR?

File: nautobot/dcim/api/serializers.py [hunk 23]
```diff
@@ -200,6 +200,21 @@ class Meta:
...
         validators = []
 
+        detail_view_config = [
+            {
+                "Location": {
+                    "fields": ["name", "location_type", "tenant", "description", "facility", "asn", "time_zone"]
+                },
+                "Contact Info": {
+                    "fields": ["physical_address", "shipping_address", "contact_name", "contact_phone", "contact_email"]
+                },
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:210]: Missing `latitude` and `longitude` fields. Is there anything we can add in the detail_view_config processing to automatically detect if fields are missing from the config?

File: nautobot/ui/src/components/ObjectTable.js [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+                const column_data =
+                    idx === 0
+                        ? props.row.original
+                        : props.row.original[props.column.id];
+                return (
+                    <ObjectTableItem
+                        name={key}
+                        obj={props.getValue()}
+                        url={getTableItemLink(idx, column_data)}
+                    />
```
Comment (glennmatthews) [inline nautobot/ui/src/components/ObjectTable.js:39]: Is the comment inaccurate here? I don't see how `column_data = {"display": "Active"}` then leads to a useful `getTableItemLink(idx, {"display": "Active"})` call when it seems like `getTableItemLink()` expects an object with an `id` or `url` key.

File: nautobot/ui/src/components/ObjectTableItem.js [hunk 25]
```diff
@@ -58,8 +58,9 @@ function TableColumnDisplay({ obj }) {
     }
 }
 
-export default function TableItem({ name, obj, url }) {
-    if (url) {
+// TODO(timizuo): create a ObjectTableItem folder and move ObjectTable, ObjectTableItem and ObjectListTable to the folder.
```
Comment (glennmatthews) [inline nautobot/ui/src/components/ObjectTableItem.js:61]: Still a TODO? If so why not do it now?

File: nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+    if (error || schemaIsError) {
+        return <div>Failed to load Change Log Data</div>;
+    }
+
+    const defaultHeaders = schema.view_options.list_display_fields;
+    const tableHeaders = schema.view_options.fields;
+    const tableData = data?.results || [];
+
+    return (
+        <NautobotGrid columns="4">
```
Comment (glennmatthews) [inline nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js:51]: why `columns="4"` when there's only a single `NautobotGridItem`?

File: nautobot/ui/src/components/Retrieve/RenderNotesTab.js [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+    if (error || schemaIsError) {
+        return <div>Failed to load Notes Data</div>;
+    }
+
+    const defaultHeaders = schema.view_options.list_display_fields;
+    const tableHeaders = schema.view_options.fields;
+    const tableData = data?.results || [];
+
+    return (
+        <NautobotGrid columns="4">
```
Comment (glennmatthews) [inline nautobot/ui/src/components/Retrieve/RenderNotesTab.js:53]: Same question here re `columns="4"` with a single NautobotGridItem?

File: nautobot/dcim/api/serializers.py (hunk 23 above)
Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py:210]: What of the user do not want to show all model fields?

File: nautobot/core/api/metadata.py (hunk 19 above)
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:89]: Nope, as ManyRelated field, would is a recursive call, and would end up getting to this point

File: nautobot/core/api/metadata.py [hunk 28]
```diff
@@ -56,10 +60,11 @@ def _get_field_properties(self, field: schema.SerializerType, name: str) -> Dict
 
         This has been overloaded with an `elif` to account for `ManyRelatedField`.
         """
-        result = {}
         type_map_obj = self._get_type_map_value(field)
-        result["type"] = type_map_obj["type"]
-        result["title"] = self._get_title(field, name)
+        result = {
+            "type": type_map_obj["type"] or "string",
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:65]: For some readonly fields, previously readonly fields where not part of schema.

File: nautobot/core/api/metadata.py (hunk 20 above)
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:103]: Its not always accurate: below response of `print("====> " + name, (type_map_obj["readOnly"], field.read_only))`

```bash
====> id (None, True)
====> object_type (None, True)
====> display (None, True)
====> url (None, True)
====> natural_key_slug (None, True)
====> tree_depth (None, True)
====> time_zone (None, False)
====> circuit_count (None, True)
====> device_count (None, True)
====> name (None, False)
====> slug (None, False)
====> description (None, False)
====> facility (None, False)
====> asn (None, False)
```

File: nautobot/core/api/metadata.py (hunk 21 above)
Comment (timizuoebideri1) [reply nautobot/core/api/metadata.py:306]: Its now `composite_key`?

File: nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js (hunk 26 above)
Comment (timizuoebideri1) [reply nautobot/ui/src/components/Retrieve/RenderChangeLogsTab.js:51]: I'm not sure why, but if I don't set the column, the table doesn't take up the entire width.

File: nautobot/core/api/metadata.py (hunk 21 above)
Comment (glennmatthews) [reply nautobot/core/api/metadata.py:306]: Yes.

File: nautobot/dcim/api/serializers.py (hunk 23 above)
Comment (glennmatthews) [reply nautobot/dcim/api/serializers.py:210]: Let's discuss this in standup tomorrow then. I think right now it's too easy to inadvertently omit fields that should be present.

File: nautobot/core/api/metadata.py (hunk 20 above)
Comment (jathanism) [reply nautobot/core/api/metadata.py:101]: I had added this and higher up being processed by `_get_type_map_value()`, which was relying on it being set in the schema somehow, which explains why it wasn't always accurate.

**Review by jathanism — CHANGES_REQUESTED** [review]
Minor asks.

File: nautobot/core/testing/api.py [hunk 29]
```diff
@@ -191,6 +191,37 @@ def test_options_object(self):
             response = self.client.options(url, **self.header)
             self.assertHttpStatus(response, status.HTTP_200_OK)
 
+        @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])
+        def test_options_object_detail_view_config(self):
```
Comment (glennmatthews) [inline nautobot/core/testing/api.py:195]: Feels like this and the above `test_options_object` could be merged into a single test?

File: nautobot/core/api/metadata.py [hunk 30]
```diff
@@ -256,11 +265,111 @@ def determine_metadata(self, request, view):
...
+            ]
+        """
+
+        # TODO(timizuo): Add a standardized way of handling `tenant` and `tags` fields, Possible should be on last items on second col.
+        fields_to_remove = ["composite_key", "url", "display", "status", "id"]
+        fields_to_add = ["id", "composite_key", "url"]
+
+        for section_idx, section in enumerate(view_config):
+            for idx, data in enumerate(section.items()):
+                _, value = data
```
Comment (gsnider2195) [inline nautobot/core/api/metadata.py:308]:
```suggestion
            for idx, value in enumerate(section.values()):
```

File: nautobot/ui/src/components/ObjectTable/index.js [hunk 31]
```diff
@@ -0,0 +1,3 @@
+export { default as ObjectListTable } from "./ObjectListTable"
+export { default as ObjectTable } from "./ObjectTable"
+export {default as ObjectTableItem } from "./ObjectTableItem"
```
Comment (gsnider2195) [inline nautobot/ui/src/components/ObjectTable/index.js:3]:
```suggestion
export { default as ObjectTableItem } from "./ObjectTableItem"
```

File: nautobot/ui/src/utils/url.js [hunk 32]
```diff
@@ -2,3 +2,15 @@
 export function uiUrl(apiUrl) {
     return apiUrl.replace("/api/", "/");
 }
+
+/* Build url form app_label, model_name and or pk */
```
Comment (gsnider2195) [inline nautobot/ui/src/utils/url.js:6]:
```suggestion
/* Build url from app_label, model_name and optional pk */
```

File: nautobot/ui/src/components/ObjectTable/ObjectListTable.js [hunk 33]
```diff
@@ -16,11 +16,12 @@ import {
...
     useToast,
 } from "@nautobot/nautobot-ui";
-import Paginator from "@components/paginator";
 import { useCallback, useMemo } from "react";
 
-import TableItem from "@components/TableItem";
-import LoadingWidget from "./LoadingWidget";
+import Paginator from "../paginator";
+import ObjectTableItem from "./ObjectTableItem";
+import LoadingWidget from "../LoadingWidget";
```
Comment (HanlinMiao) [inline nautobot/ui/src/components/ObjectTable/ObjectListTable.js:23]: Probably should alpha-sort these

File: nautobot/ui/src/utils/url.js [hunk 34]
```diff
@@ -2,3 +2,15 @@
 export function uiUrl(apiUrl) {
     return apiUrl.replace("/api/", "/");
 }
+
+/* Build url from app_label, model_name and optional pk */
+export function buildUrl(appLabel, modelName, pk = null, isPlugin = false) {
```
Comment (HanlinMiao) [inline nautobot/ui/src/utils/url.js:7]: Should this function account for the few `slugs` that still exist in the url after https://github.com/nautobot/nautobot/pull/3914?

**Review by HanlinMiao — APPROVED** [review]
I pulled it down and it looks noice! Just some minor changes.

File: nautobot/ui/src/components/Retrieve/RenderCol.js
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/ui/src/components/Retrieve/RenderCol.js:1]: 🦥

**Review by jathanism — APPROVED** [review]
I'm good to let this go, I just want to think about readability and consistency. :)

File: nautobot/core/testing/api.py [hunk 35]
```diff
@@ -191,6 +191,37 @@ def test_options_object(self):
...
+                    # Assert specific fields(`id`, `composite_key`, `url`) are present only at the end of the first col group fields.
+                    special_fields = ["id", "composite_key", "url"]
+                    for col_idx, col in enumerate(response_view_config):
+                        for group_idx, (group_title, group) in enumerate(col.items()):
+                            group_fields = group["fields"]
+                            # Config set on the serializer
+                            fields = detail_view_config[col_idx][group_title]["fields"]
+                            if group_idx == 0 == col_idx:
+                                self.assertEqual(special_fields, group_fields[-3:])
+                                self.assertFalse(any(field in special_fields for field in group_fields[:-3]))
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/testing/api.py]: I'm asserting that special_fields only exists at the end of the group fields and that it doesn't exist anywhere else in the group fields.

Group fields are the serializer fields that should be present in a particular column group

File: nautobot/ui/src/utils/url.js (hunk 34 above)
Comment (timizuoebideri1) [reply nautobot/ui/src/utils/url.js:7]: Not really, if a slug is passed in to the buildUrl it would still work as expected.
Do you have any in mind, just to verify

**Review by bryanculver — APPROVED** [review]

File: nautobot/docs/development/react-ui.md [hunk 36]
```diff
@@ -44,6 +44,38 @@ The libraries and tools used to build the React UI are:
...
+                    "Group Name 3": {"fields": [...]},
+                    "Group Name 4": {"fields": [...]},
+                },
+            ],
+            "include_others": False
+        }
+```
+
+In the above example, we add the `detail_view_config` attribute to the Serializer's inner `Meta` class. The value of this attribute is a dict containing `layout` and `include_others`(optional). the `layout` is a list containing two dictionaries, each representing the two columns of the detail view. The first dictionary represents the fields in the first column, while the second dictionary represents the fields in the second column. Each dictionary consists of a key-value pair, where the key is the name of the grouping, and the value is a list of the model fields that should be included in that grouping.
+The optional key `include_others`, when set to `True`, adds missing serializer fields from the `detail_view_config["layout"]` to the view_config layout.
```
Comment (glennmatthews) [inline nautobot/docs/development/react-ui.md:72]: I'd still like us to revisit this in the near future - my feeling is the number of cases where we want to *omit* fields from the UI is the minority, and in the vast majority of cases we'll want every serializer field to be represented in the UI.

File: nautobot/ui/src/components/__tests__/Table.test.js [hunk 37]
```diff
@@ -1,36 +1,39 @@
...
+// const mockHeaders = [
+//     { key: "name", dataIndex: "name", title: "Name" },
+//     { key: "description", dataIndex: "description", title: "Description" },
+// ];
+// const mockData = [
+//     { id: 1, name: "Item 1", description: "Description 1" },
+//     { id: 2, name: "Item 2", description: "Description 2" },
+// ];
 
+// FixMe(timizuo): Cannot test ObjectTable as it implements @nautobot-ui TableRenderer component which
```
Comment (glennmatthews) [inline nautobot/ui/src/components/__tests__/Table.test.js:14]: Is this test going to be addressed in a follow-on issue, or as part of this PR?

**Review by glennmatthews — APPROVED** [review]
I still have some reservations, but let's get this merged and iterate on it.
